src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -1251,6 +1282,9 @@ bool PartMergerWriter::mutateOriginalPartAndPrepareProjections()
 
         ctx->out->write(cur_block);
 
+        if (ctx->has_lightweight_delete)
+            existing_rows_count += MutationHelpers::getExistingRowsCount(cur_block);
```

Comment:
As I understand this calculation should be performed only if setting `exclude_deleted_rows_for_part_size_in_merge` is enabled, otherwise it will not be used and it's a waste of time. Also, doing this here is not very efficient - best way it should be done as part of `DELETE FROM` mutation, though it may be very cumbersome, so we can leave it here for now but add a comment something like `TODO: move this calculation to DELETE FROM mutation`

Reply:
Let me add the setting check here.
As for the second suggestion, actually nothing happens in `DELETE FROM` itself, it was transformed into `ALTER {table} UPDATE _row_exists = 0` mutation.

Reply:
>it was transformed into ALTER {table} UPDATE _row_exists = 0 mutation

yes, exactly - it's why it would be cumbersome to implement it - will require manipulations with qury pipline I think...

Reply:
I see, you mean getting the affected rows count before mutation executes, looks difficult to implement.
I added the `TODO` comment for now